File: runtime/typescript/packages/core/src/harness/replay-verifier.ts
// Copyright (c) Microsoft. All rights reserved.

import {
  ReplayMismatch,
  ReplayVerificationRequest,
  ReplayVerificationResult,
  type ReplayJournalRecord,
} from "../model/index.js";

export { ReplayVerificationRequest, ReplayVerificationResult };

function comparable(record: ReplayJournalRecord | undefined): string | undefined {
  return record === undefined ? undefined : JSON.stringify(record.save());
}

/** Verifies an actual normalized replay journal against expected records. */
export class ReferenceReplayVerifier {
  verify(request: ReplayVerificationRequest): ReplayVerificationResult {
    const expected = request.expected ?? [];
    const actual = request.actual ?? [];
    const max = Math.max(expected.length, actual.length);
    const mismatches: ReplayMismatch[] = [];

    for (let index = 0; index < max; index += 1) {
      const expectedRecord = expected[index];
      const actualRecord = actual[index];
      if (comparable(expectedRecord) !== comparable(actualRecord)) {
        mismatches.push(
          new ReplayMismatch({
            index,
            expected: expectedRecord,
            actual: actualRecord,
            message: expectedRecord === undefined
              ? "Unexpected extra replay record"
              : actualRecord === undefined
                ? "Missing replay record"
                : "Replay record mismatch",
          }),
        );
      }
    }

    return new ReplayVerificationResult({
      status: mismatches.length === 0 ? "passed" : "failed",
      mismatches,
      expectedCount: expected.length,
      actualCount: actual.length,
    });
  }
}

File: runtime/csharp/Prompty.Core/ReplayVerifier.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Text.Json;

namespace Prompty.Core;

public sealed class ReferenceReplayVerifier
{
    public ReplayVerificationResult Verify(ReplayVerificationRequest request)
    {
        var expected = request.Expected ?? [];
        var actual = request.Actual ?? [];
        var max = Math.Max(expected.Count, actual.Count);
        var mismatches = new List<ReplayMismatch>();

        for (var index = 0; index < max; index++)
        {
            var expectedRecord = index < expected.Count ? expected[index] : null;
            var actualRecord = index < actual.Count ? actual[index] : null;
            if (Comparable(expectedRecord) != Comparable(actualRecord))
            {
                mismatches.Add(new ReplayMismatch
                {
                    Index = index,
                    Expected = expectedRecord,
                    Actual = actualRecord,
                    Message = expectedRecord is null
                        ? "Unexpected extra replay record"
                        : actualRecord is null
                            ? "Missing replay record"
                            : "Replay record mismatch"
                });
            }
        }

        return new ReplayVerificationResult
        {
            Status = mismatches.Count == 0 ? ReplayVerificationStatus.Passed : ReplayVerificationStatus.Failed,
            Mismatches = mismatches,
            ExpectedCount = expected.Count,
            ActualCount = actual.Count
        };
    }

    private static string? Comparable(ReplayJournalRecord? record)
    {
        return record is null ? null : JsonSerializer.Serialize(record.Save());
    }
}
